sdk/inkrecognizer/src/InkRecognitionUnitBuilder.ts

```diff
@@ -0,0 +1,368 @@
+      result.lines.push(unit as Line);
+      break;
+    case 'paragraph':
+      unit = createParagraph(json);
+      result.paragraphs.push(unit as Paragraph);
+      break;
+    case 'writingRegion':
+      unit = createWritingRegion(json);
+      result.writingRegions.push(unit as WritingRegion);
+      break;
+    case 'inkBullet':
+      unit = createInkBullet(json as JsonInkBullet);
+      result.inkBullets.push(unit as InkBullet);
+      break;
+    default:
+      unit = createInkRecognitionUnit(json);
+      break;
+  }
+  return unit;
+}
+
+export function parseInkRecognitionResult(responseText: string): InkRecognitionResult {
+  let idToRecognitionUnitMap: Map<Number, InkRecognitionUnit> = new Map<Number, InkRecognitionUnit>();
+  let idToJsonRecognitionUnitMap: Map<Number, JsonInkRecognitionUnit> = new Map<Number, JsonInkRecognitionUnit>();
+
+  let result = {} as InkRecognitionResult;
+  result.inkDrawings = [];
+  result.inkWords = [];
+  result.inkBullets = [];
+  result.lines = [];
+  result.paragraphs = [];
+  result.writingRegions = [];
+
+  // Parse InkRecognitionUnits from json
+  const json: JsonInkRecognitionResult = JSON.parse(responseText) as JsonInkRecognitionResult;
+  json.recognitionUnits.forEach((jsonUnit: JsonInkRecognitionUnit) => {
+    let unit = processInkRecognitionUnit(jsonUnit, result);
+    idToRecognitionUnitMap.set(unit.id, unit);
+    idToJsonRecognitionUnitMap.set(unit.id, jsonUnit);
+  });
+
+  // Fill parent and children recognitionUnits
+  idToRecognitionUnitMap.forEach((value: InkRecognitionUnit, key: Number) => {
+    if (idToJsonRecognitionUnitMap.has(key)) {
+      let jsonUnit = idToJsonRecognitionUnitMap.get(key);
+      if (jsonUnit) {
+        let parentId = jsonUnit.parentId;
+        if (idToRecognitionUnitMap.has(parentId)) {
+          let parent = idToRecognitionUnitMap.get(parentId);
+          if (parent) {
+            value.parent = parent;
+          }
+        }
+
+        let childIds = jsonUnit.childIds;
+        if (childIds) {
+          childIds.forEach((cid) => {
+            if (idToRecognitionUnitMap.has(cid)) {
+              let child = idToRecognitionUnitMap.get(cid);
+              if (child) {
+                value.children.push(child);
+              }
+            }
+          });
+        }
+      }
+    }
+  });
+
+  // Merge properties from Root
+  const InkRecognitionRootId = 0;
+  result.id = InkRecognitionRootId;
+  result.kind = InkRecognitionUnitKind.Root;
+  result.parent = {} as InkRecognitionUnit;
+  result.children = [];
+  result.strokeIds = [];
+  result.boundingRectangle = {} as Rectangle;
+  result.rotatedBoundingRectangle = [];
+
+  if (!idToRecognitionUnitMap.has(InkRecognitionRootId)) {
+    // InkRecognitionRoot will be available in GA, build here before then.
+    // throw new Error("Fail to pass response: no InkRecognitionRoot present.");
+    result.inkDrawings.forEach(drawing => { result.children.push(drawing); });
```

Comment:
Let me know if you like this: `result.children.push(...result.inkDrawings);`. Feel free to discard this suggestion if you don't like it!

Reply:
Thanks for suggestion, this is good to learn (I am quite new with typescript;) )

Reply:
`result.children = result.children.concat(result.inkDrawings)` would be a more efficient and straight forward way to do this imo.

Reply:
will use _concat_ if that's more efficient.